Update naming-rules.md (#21470)
dotnet/docs: docs/fundamentals/code-analysis/style-rules/naming-rules.md

Naming rules concern the naming of .NET programming language code elements, such as classes, properties, and methods. For example, you can specify that public members must be capitalized or that private fields must begin with `_`.

A naming rule has three parts:

> * the symbol group -- to which symbols does the rule apply?

Author: I'm a little wary of creating this new term -- **symbol group**. But if you feel it's self-evident enough, I'll accept this.

* what naming style to associate with the rule, and
* the severity for enforcing the convention.

> 

Reviewer: ```suggestion

You define naming rules in an EditorConfig file.

```

## General syntax

> To define a naming rule, symbol group, or naming style, write one or more property settings with the following syntax:

Reviewer: ```suggestion
To define a naming rule, symbol group, or naming style, add one or more property settings using the following syntax:
```

Author: I think _add_ implies that some part of the definition exists somewhere else, which isn't true -- once you have a property setting, you have a definition.

Reviewer: Then I'd suggest "...define one or more...". "Write" is not a term we generally use in the docs to mean that.

Author: Would "... set one or more properties using the following syntax:" be better?

Reviewer: That sounds good!

```none
<prefix>.<title>.<propertyName> = <propertyValue>
```

> Each property should only be set once, but some settings allow multiple comma-separated values.

Reviewer: ```suggestion
Each property should only be set once, but some settings allow multiple, comma-separated values.
```

The order of the properties is not important.

### \<prefix>

> **\<prefix>** specifies which kind of element is being defined -- naming rule, symbol group, or naming style -- and must be one of the following:

Reviewer: ```suggestion
**\<prefix>** specifies which kind of element is being defined&mdash;naming rule, symbol group, or naming style&mdash;and must be one of the following:
```

| To set a property for | Use the prefix | Example |
| --- | --- | -- |
| Symbol group | `dotnet_naming_symbols` | `dotnet_naming_symbols.interface.applicable_kinds = interface` |
| Naming style | `dotnet_naming_style` | `dotnet_naming_style.pascal_case.capitalization = pascal_case` |

> | Naming rule | `dotnet_naming_rule` | `dotnet_naming_rule.types_should_be_pascal_case.severity = suggestion`

Reviewer: Can you move this up to be the first row in the table, to match the order you present the properties on line 44.

Each type of definition -- [naming rule](#naming-rule-properties), [symbol group](#symbol-group-properties), or [naming style](#style-properties)  -- has its own supported properties, as described below.

### \<title>

> **\<title>** is a descriptive name that you choose, which associates multiple property settings into a single definition. For example, the following:

Reviewer: ```suggestion
**\<title>** is a descriptive name that you choose that associates multiple property settings into a single definition. For example, the following properties produce two naming rule definitions, `interface` and `types`, each of which has two properties set on it.
```

Author: 1. Is there any way to avoid the double _that_ of _a descriptive name that you choose that associates_?
2. I think it important to differentiate between "naming rule definition" and "symbol group definition". I'll address this in a later commit.

```ini
dotnet_naming_symbols.interface.applicable_kinds = interface
dotnet_naming_symbols.interface.applicable_accessibilities = public, internal, private, protected, protected_internal, private_protected

dotnet_naming_symbols.types.applicable_kinds = class, struct, interface, enum, delegate
dotnet_naming_symbols.types.applicable_accessibilities = public, internal, private, protected, protected_internal, private_protected
```

> produces two symbol group definitions -- `interface` and `types` -- each of which has two properties set on it.

Reviewer: ```suggestion
```

## Naming rule properties

> All naming rule properties are required.

Reviewer: ```suggestion
All naming rule properties are required for a rule to take effect.
```

| Property | Description |
| -- | -- |

> | `symbols` | The title of the symbol group, defining the symbols to which this rule should be applied |

Reviewer: ```suggestion
| One of the [symbol group properties](#symbol-group-properties) | Defines the symbols to which this rule should be applied, for example, kind, accessibility, and modifiers. |
```

Author: This isn't correct. We're listing the properties which can be set on a naming rule definition; `symbols` is one of those properties, and the value will be the title of one of the symbol groups defined elsewhere. This is entirely distinct from the properties that can set on a symbol group.

> | `style` | The title of the naming style which should be associated with this rule |

Reviewer: ```suggestion
| One of the [naming style properties](#naming-style-properties) | Describes the naming style to associate with this rule, for example, required prefix or suffix. |
```

Author: Again, we're listing the properties which can be set for a naming rule. The value for `style` will be a style title used elsewhere.

For example, once I have set a property on the style like this:
```
dotnet_naming_style.begins_with_i.required_prefix = I
dotnet_naming_style.begins_with_i.capitalization = pascal_case
```
I now have a style definition with the title `begins_with_i`. I can then associate that style with a given naming rule:, by setting the naming rule's `style` property:
```
dotnet_naming_rule.interface_should_be_begins_with_i.style = begins_with_i
dotnet_naming_rule.interface_should_be_begins_with_i.symbols = interface
dotnet_naming_rule.interface_should_be_begins_with_i.severity = suggestion
```
In this case, the `interface_should_be_begins_wth_i` naming rule.

The `symbols` property works similarly: I define a symbol group by setting one (or more) of the symbol group properties with a common title. I can then set the naming rule's `symbols` property to that common title.

> | `severity` |  Any of the available [severity levels](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/configuration-options#severity-level)<sup>1</sup> |

Reviewer: ```suggestion
| `severity` |  Sets the severity with which to enforce the naming rule. Set the associated value to one of the available [severity levels](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/configuration-options#severity-level).<sup>1</sup> |
```

| `applicable_accessibilities` | Accessibility levels of the symbols in the group | `*` (use this value to specify all accessibility levels)<br/>`public`<br/>`internal` or `friend`<br/>`private`<br/>`protected`<br/>`protected_internal` or `protected_friend`<br/>`private_protected`<br/>`local` (for symbols defined within a method) | Yes |
| `required_modifiers` | Only match symbols with _all_ the specified modifiers <sup>2</sup> | `abstract` or `must_inherit`<br/>`async`<br/>`const`<br/>`readonly`<br/>`static` or `shared` <sup>3</sup> | No |

**Notes:**

1. Tuple members aren't currently supported in `applicable_kinds`.
2. The symbol group matches _all_ the modifiers in the `required_modifiers` property.  If you omit this property, no specific modifiers are required for a match. This means a symbol's modifiers have no effect on whether or not this rule is applied.

> 3. If your group has `static` or `shared` in the `required_modifiers` property, the group will also include `const` symbols because they are implicitly `static`/`Shared`. However, you can create a new naming rule with its' own symbol group of `const` symbols.

Reviewer: ```suggestion
3. If your group has `static` or `shared` in the `required_modifiers` property, the group will also include `const` symbols because they are implicitly `static`/`Shared`. However, If you don't want the `static` naming rule to apply to `const` symbols, you can create a new naming rule with a symbol group of `const`.
```

## Naming style properties

> A naming style defines the conventions we want to enforce with a naming rule. For example:

Reviewer: ```suggestion
A naming style defines the conventions you want to enforce with the rule. For example:
```

1. Tuple members aren't currently supported in `applicable_kinds`.
2. The symbol group matches _all_ the modifiers in the `required_modifiers` property.  If you omit this property, no specific modifiers are required for a match. This means a symbol's modifiers have no effect on whether or not this rule is applied.

> 3. If your group has `static` or `shared` in the `required_modifiers` property, the group will also include `const` symbols because they are implicitly `static`/`Shared`. However, If you don't want the `static` naming rule to apply to `const` symbols, you can create a new naming rule with a symbol group of `const`.

Reviewer: ```suggestion
3. If your group has `static` or `shared` in the `required_modifiers` property, the group will also include `const` symbols because they are implicitly `static`/`Shared`. However, if you don't want the `static` naming rule to apply to `const` symbols, you can create a new naming rule with a symbol group of `const`.
```

The following *.editorconfig* file contains a naming convention that specifies that public properties, methods, fields, events, and delegates must be capitalized. Notice that this naming convention specifies multiple kinds of symbol to apply the rule to, using a comma to separate the values.

> ```ini

Author: In order to further emphasize the difference between "setting a property which defines a symbol group or naming style" and "setting the naming rule property to the title of another element", I've reordered the property settings in the final example:

* first, the symbol group definition,
* then, the naming style definition,
* and finally, the naming rule definition, which includes setting the symbol group and naming style for the naming rule.
